packages/cli/src/services/project.service.ts

```diff
@@ -21,6 +24,81 @@ export class ProjectService {
 		private readonly roleService: RoleService,
 	) {}
 
+	private get workflowService() {
+		return import('@/workflows/workflow.service').then(({ WorkflowService }) =>
+			Container.get(WorkflowService),
+		);
+	}
+
+	private get credentialsService() {
+		return import('@/credentials/credentials.service').then(({ CredentialsService }) =>
+			Container.get(CredentialsService),
+		);
+	}
+
+	private get activeWorkflowRunner() {
+		return import('@/ActiveWorkflowRunner').then(({ ActiveWorkflowRunner }) =>
+			Container.get(ActiveWorkflowRunner),
+		);
+	}
+
+	async deleteProject(user: User, projectId: string) {
+		const workflowService = await this.workflowService;
+		const credentialsService = await this.credentialsService;
+		const activeWorkflowRunner = await this.activeWorkflowRunner;
+
+		await this.projectRelationRepository.manager.transaction(async (em) => {
+			const project = await this.getProjectWithScope(user, projectId, 'project:delete', em);
+
+			if (!project) {
+				throw new NotFoundError(`Could not find project with ID: ${projectId}`);
+			}
+
+			// 0. check if this is a team project
+			if (project.type !== 'team') {
+				throw new ForbiddenError(
+					`Can't delete project. Project with ID "${projectId}" is not a team project.`,
+				);
+			}
+
+			// 1. delete workflows owned by this project
+			const ownedSharedWorkflows = await em.find(SharedWorkflow, {
+				where: { projectId: project.id, role: 'workflow:owner' },
+				relations: { workflow: true },
+			});
+
+			for (const sharedWorkflow of ownedSharedWorkflows) {
+				// NOTE: This is supposed to happen outside the transaction.
+				// We'd otherwise need to pass the em through to a lot of functions even ending up in core.
+				// See: https://github.com/n8n-io/n8n/pull/8904#discussion_r1530150510
+				await activeWorkflowRunner.remove(sharedWorkflow.workflow.id);
+				await workflowService.deleteInactiveWorkflow(user, sharedWorkflow.workflow, em);
+			}
```

Comment:
I'd suggest moving this code to outside the transaction and only begin the transaction for the actual removal wdyt?

This way, if something fails while deactivating we abort prematurely and warn the user, and we still have the safety of the transaction.

Does this sound too crazy?

Worst case scenario we end up in a situation where the workflow was deactivated but a few remain active, which should not be too bad.

Reply:
I can, but how do I do that without moving this out as well?

```ts
			const ownedSharedWorkflows = await em.find(SharedWorkflow, {
				where: { projectId: project.id, role: 'workflow:owner' },
				relations: { workflow: true },
			});
```

And if I move this out we have another race condition. A new workflow could be created while the project is being deleted.

Reply:
I mean I could do the `find` twice and abort the transaction if I get different results in between.

I'm not happy with how much context a future reader will need to understand all the intricacies of this.

Reply:
Nevermind, `find`s don't lock the tables, so the problem that a new workflow can be added while the project is being deleted is something that would happen even with the `find` being in the transaction.